scripts/local_cloudbuild.py

```diff
@@ -0,0 +1,322 @@
+import re
+import shlex
+import subprocess
+import sys
+
+import yaml
+
+
+# Exclude non-printable control characters (including newlines)
+PRINTABLE_REGEX = re.compile(r"""^[^\x00-\x1f]*$""")
+
+# File template
+BUILD_SCRIPT_HEADER = """\
+#!/bin/bash
+# This is a generated file.  Do not edit.
+
+set -euo pipefail
+
+SOURCE_DIR=.
+
+# Setup staging directory
+HOST_WORKSPACE=$(mktemp -d)
+function cleanup {
+    if [ "${HOST_WORKSPACE}" != '/' -a -d "${HOST_WORKSPACE}" ]; then
+        rm -rf "${HOST_WORKSPACE}"
+    fi
+}
+trap cleanup EXIT
+
+# Copy source to staging directory
+echo "Copying source to staging directory ${HOST_WORKSPACE}"
+rsync -avzq --exclude=.git "${SOURCE_DIR}" "${HOST_WORKSPACE}"
+
+# Build commands
+"""
+
+BUILD_SCRIPT_FOOTER = """\
+# End of build commands
+
+echo "Build completed successfully"
+"""
+
+
+# Validated cloudbuild recipe + flags
+CloudBuild = collections.namedtuple('CloudBuild', 'output_script run steps')
+
+# Single validated step in a cloudbuild recipe
+Step = collections.namedtuple('Step', 'args dir_ env name')
+
+
+def get_field_value(container, field_name, field_type):
+    """Fetch a field from a container with typechecking and default values.
+
+    The field value is coerced to the desired type.  If the field is
+    not present, a instance of `field_type` is constructed with no
+    arguments and used as the default value.
+
+    Args:
+        container (dict): Object decoded from yaml
+        field_name (str): Field that should be present in `container`
+        field_type (type): Expected type for field value
+
+    Returns:
+        any: Fetched or default value of field
+
+    Raises:
+        ValueError: if field value cannot be converted to the desired type
+    """
+    try:
+        value = container[field_name]
+    except (IndexError, KeyError):
+        return field_type()
+
+    msg = 'Expected "{}" field to be of type "{}", but found type "{}"'
+    if not isinstance(value, field_type):
+        # list('some string') is a successful type cast as far as Python
+        # is concerned, but doesn't exactly produce the results we want.
+        # We have a whitelist of conversions we will attempt.
+        whitelist = (
+            (float, str),
+            (int, str),
+            (str, float),
+            (str, int),
+            (int, float),
+            )
+        if (type(value), field_type) not in whitelist:
+            raise ValueError(msg.format(field_name, field_type, type(value)))
+
+    try:
+        value = field_type(value)
+    except ValueError as e:
+        e.message = msg.format(field_name, field_type, type(value))
+        raise
+    return value
+
+
+def get_cloudbuild(raw_config, args):
+    """Read and validate a cloudbuild recipe
+
+    Args:
+        raw_config (dict): deserialized cloudbuild.yaml
+        args (argparse.Namespace): ccommand line flags
+
+    Returns:
+        CloudBuild: valid configuration
+    """
+    if not isinstance(raw_config, dict):
```

Comment:
I have to disagree with you on this.  This is coming (via yaml parse) from user input, thus the principle of "minimize attack surface" and "trust no one" is paramount.  Basically reverse Postel's Law: Allow the least you possibly can.  In this code specifically, we know the structure coming from yaml parsing (or possibly json later), so if it's a mapping type, it's a dict.